language/move-lang/src/unit_test/filter_test_members.rs

```diff
@@ -129,6 +131,70 @@ fn filter_tests_from_module(
     })
 }
 
+/// If a module is being compiled in test mode, this inserts a function that calls a native
+/// function `0x1::UnitTest::create_signers_for_testing` that only exists if the VM is being run with the
+/// "unit_test" feature flag set. This will then cause the module to fail to link if an attempt is
+/// made to publish a module that has been compiled in test mode on a VM that is not running in
+/// test mode.
+fn insert_test_poison(context: &mut Context, mloc: Loc, members: &mut Vec<P::ModuleMember>) {
+    if !context.env.flags().is_testing() {
+        return;
+    }
+
+    let signature = P::FunctionSignature {
+        type_parameters: vec![],
+        parameters: vec![],
+        return_type: sp(mloc, P::Type_::Unit),
+    };
+
+    let module_access = P::ModuleIdent {
+        locs: (mloc, mloc),
+        value: (Address::DIEM_CORE, "UnitTest".to_string()),
```

Comment:
Can you gate this that the module actually exists? I really don't want references to `DIEM_CORE` floating around in testing code. 

Maybe check for a module called `Std::UnitTest` existing and error if it doesnt?

Reply:
Then again, I suppose you will get a very interesting error message if it doesn't exist...

Reply:
I think checking to make sure that this module exists in one of the deps or sources makes sense. You will indeed get some interesting error messages if it doesn't exist, but that isn't great...much better to have a proper error message for that case.

 I'm not sure how to check for the `Std::` part, is this incoming with the Named Addresses PR? If so I can wait to land this until that PR gets in and then rebase and refactor on top of that.

Reply:
We will have to sort out the discussion of the naming and things, but I'd be a lot happier about "Std" being in the compiler than "DIEM_CORE"